File: src/main/kotlin/dev/jefy/connectpro/user/domain/service/OtpCodeGenerator.java
package dev.jefy.connectpro.user.domain.service;


import org.springframework.stereotype.Service;

import java.time.Instant;

import dev.jefy.connectpro.user.domain.repository.TokenRepository;
import dev.jefy.connectpro.user.domain.vo.OtpCode;
import lombok.RequiredArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Service
@RequiredArgsConstructor
public class OtpCodeGenerator {
    private final TokenRepository tokenRepository;

    public OtpCode generate() {
        OtpCode code;
        do {
            code = OtpCode.generate();
        } while (tokenRepository.isCodeActiveAndNotValidated(code, Instant.now()));
        return code;
    }
}

File: src/main/kotlin/dev/jefy/connectpro/user/domain/service/OtpCodeGenerator.kt
package dev.jefy.connectpro.user.domain.service

import org.springframework.stereotype.Service
import dev.jefy.connectpro.user.domain.repository.TokenRepository
import dev.jefy.connectpro.user.domain.vo.OtpCode
import java.time.Instant

@Service
class OtpCodeGenerator(private val tokenRepository: TokenRepository) {

    fun generate(): OtpCode {
        var code: OtpCode
        do {
            code = OtpCode.generate()
        } while (tokenRepository.isCodeActiveAndNotValidated(code, Instant.now()))
        return code
    }
}